dataloaders/helpers.py

```diff
@@ -78,17 +86,21 @@ def dataloader_from_dataset(
     def collate_fn(batch):
         """`batch` is a list with `batch_size` elements. Each element is a dict with
         `label` and `text_ids` keys."""
-        d = collections.defaultdict(list)
-        for el in batch:
-            for k in text_columns:
-                d[k].append(torch.stack(el[k])) # shape (seq_length, 50)
-            for k in label_columns:
-                d[k].append(el[k]) # shape ()
-        # stack all lists into tensors and return in the proper order
-        return tuple(torch.stack(d[k]) for k in text_columns) + tuple(torch.stack(d[k]).float() for k in label_columns)
+        batch = batch[0] # RandomSampler wraps batch in extra dimension
+        return (
+            tuple(torch.tensor(batch[k]) for k in text_columns) 
+          + tuple(torch.tensor(batch[k]).float() for k in label_columns)
+        )
+    
+    # epochs = 5
+    sampler = BatchSampler(
+        RandomSampler(dataset, replacement=True, num_samples=len(dataset)),
+        batch_size = batch_size, drop_last = drop_last
+    )
```

Comment:
I read the [PyTorch Forum post](https://discuss.pytorch.org/t/observing-strange-loss-jumps-between-epochs/64066/16), and just to confirm, using `shuffle=True` effectively does the same thing, but using `BatchSampler(RandomSampler(...))` makes the sawtooth shape smooth, correct? 

That's how I understood part of [K. Frank's reply](https://discuss.pytorch.org/t/observing-strange-loss-jumps-between-epochs/64066/13), copied here:

> I’m inclined to doubt that there is anything wrong with the
> DataLoader (other than that it’s doing what it was designed
> to do – which may or may not be what you want).
> 
> The DataLoader gives you an entire epoch of samples, one batch
> at a time, without any repeats. When you go on to a new epoch,
> you sometimes will get repeats of recently seen samples. You
> would expect this to give the network an advantage, and lead to
> a lower loss. The only question is whether this effect would be
> large enough to be significant and/or noticeable. (I suppose
> you could call this a “statistical bias” of a sort, but it is what the
> DataLoader is supposed to do.)

Reply:
Also, given commit [`disable sampling w/ replacement`](https://github.com/jscuds/rf-bert/pull/25/commits/5a2611abe45b9ce524ca8fee7e439babefc6377e), it seems like this is no longer an issue after fixing the model.